# Start

In [63]:
import pandas as pd

In [64]:
import sys
sys.path.append("..")

from scripts.helper_funcs import get_encoding

In [65]:
loc = "../data/raw/data.csv"

encoding = get_encoding(loc)

print(encoding)

ascii


In [66]:
df = pd.read_csv(loc)

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

In [68]:
df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [69]:
df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

In [70]:
df.head(5)

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [71]:
df = df.rename(columns={"Make": "make", "Model": "model", "Year": "year","Engine Fuel Type": "engine_fuel_type", "Engine HP": "engine_hp", "Engine Cylinders": "engine_cylinders", "Transmission Type": "transmission_type", "Driven_Wheels": "driven_wheels", "Number of Doors": "number_of_doors", "Market Category": "market_category", "Vehicle Size": "vehicle_size", "Vehicle Style": "vehicle_style", "highway MPG": "highway_mpg", "city mpg": "city_mpg", "Popularity": "popularity", "MSRP": "msrp"})
df

make       model  year                engine_fuel_type  engine_hp  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       engine_cylinders transmission_type      driven_wheels  number_of_doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             market_category vehicle_size  vehicle_style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway_mpg  city_mpg  popularity   msrp  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

# Identifiying duplicates

In [72]:
df[df.duplicated()]

make     model  year             engine_fuel_type  engine_hp  \
14        BMW  1 Series  2013  premium unleaded (required)      230.0   
18       Audi       100  1992             regular unleaded      172.0   
20       Audi       100  1992             regular unleaded      172.0   
24       Audi       100  1993             regular unleaded      172.0   
25       Audi       100  1993             regular unleaded      172.0   
...       ...       ...   ...                          ...        ...   
11481  Suzuki      X-90  1998             regular unleaded       95.0   
11603   Volvo      XC60  2017             regular unleaded      302.0   
11604   Volvo      XC60  2017             regular unleaded      240.0   
11708  Suzuki       XL7  2008             regular unleaded      252.0   
11717  Suzuki       XL7  2008             regular unleaded      252.0   

       engine_cylinders transmission_type      driven_wheels  number_of_doors  \
14                  6.0            MANUAL   rear wheel drive              2.0   
18                  6.0            MANUAL  front wheel drive              4.0   
20                  6.0            MANUAL  front wheel drive              4.0   
24                  6.0            MANUAL  front wheel drive              4.0   
25                  6.0            MANUAL  front wheel drive              4.0   
...                 ...               ...                ...              ...   
11481               4.0            MANUAL   four wheel drive              2.0   
11603               4.0         AUTOMATIC    all wheel drive              4.0   
11604               4.0         AUTOMATIC  front wheel drive              4.0   
11708               6.0         AUTOMATIC    all wheel drive              4.0   
11717               6.0         AUTOMATIC  front wheel drive              4.0   

                    market_category vehicle_size vehicle_style  highway_mpg  \
14               Luxury,Performance      Compact         Coupe           28   
18                           Luxury      Midsize         Sedan           24   
20                           Luxury      Midsize         Sedan           24   
24                           Luxury      Midsize         Sedan           24   
25                           Luxury      Midsize         Sedan           24   
...                             ...          ...           ...          ...   
11481                           NaN      Compact       2dr SUV           26   
11603  Crossover,Luxury,Performance      Midsize       4dr SUV           29   
11604              Crossover,Luxury      Midsize       4dr SUV           30   
11708                     Crossover      Midsize       4dr SUV           22   
11717                     Crossover      Midsize       4dr SUV           22   

       city_mpg  popularity   msrp  
14           19        3916  31500  
18           17        3105   2000  
20           17        3105   2000  
24           17        3105   2000  
25           17        3105   2000  
...         ...         ...    ...  
11481        22         481   2000  
11603        20         870  46350  
11604        23         870  40950  
11708        15         481  29149  
11717        16         481  27499  

[715 rows x 16 columns]

In [73]:
df[df.duplicated()].isna().sum()

make                   0
model                  0
year                   0
engine_fuel_type       0
engine_hp              0
engine_cylinders       0
transmission_type      0
driven_wheels          0
number_of_doors        0
market_category      366
vehicle_size           0
vehicle_style          0
highway_mpg            0
city_mpg               0
popularity             0
msrp                   0
dtype: int64

## Missing Values

- engine_fuel_type 3
- engine_hp 69
- engine_cylinders 30 
- market_category 3742

## Duplicates

- 715 rows × 16 columns

In [74]:
df_no_duplicates = df.drop_duplicates()

In [75]:
df_no_duplicates["engine_cylinders"].unique()

array([ 6.,  4.,  5.,  8., 12.,  0., nan, 10.,  3., 16.])

In [76]:
df_no_duplicates[df_no_duplicates["engine_cylinders"].isna()]

make    model  year             engine_fuel_type  engine_hp  \
1983   Chevrolet  Bolt EV  2017                     electric      200.0   
1984   Chevrolet  Bolt EV  2017                     electric      200.0   
3716  Volkswagen   e-Golf  2015                     electric      115.0   
3717  Volkswagen   e-Golf  2015                     electric      115.0   
3718  Volkswagen   e-Golf  2016                     electric      115.0   
3719  Volkswagen   e-Golf  2016                     electric      115.0   
5778  Mitsubishi   i-MiEV  2014                     electric        NaN   
5779  Mitsubishi   i-MiEV  2016                     electric       66.0   
5780  Mitsubishi   i-MiEV  2017                     electric       66.0   
8373      Toyota  RAV4 EV  2012                     electric      154.0   
8695       Mazda     RX-7  1993             regular unleaded      255.0   
8696       Mazda     RX-7  1994             regular unleaded      255.0   
8697       Mazda     RX-7  1995             regular unleaded      255.0   
8698       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8699       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8700       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8701       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8702       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8703       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8704       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8705       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8706       Mazda     RX-8  2010  premium unleaded (required)      212.0   
8707       Mazda     RX-8  2010  premium unleaded (required)      212.0   
8708       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8709       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8710       Mazda     RX-8  2011  premium unleaded (required)      212.0   
8711       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8712       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8713       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8714       Mazda     RX-8  2011  premium unleaded (required)      212.0   

      engine_cylinders transmission_type      driven_wheels  number_of_doors  \
1983               NaN      DIRECT_DRIVE  front wheel drive              4.0   
1984               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3716               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3717               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3718               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3719               NaN      DIRECT_DRIVE  front wheel drive              4.0   
5778               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5779               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5780               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
8373               NaN      DIRECT_DRIVE  front wheel drive              4.0   
8695               NaN            MANUAL   rear wheel drive              2.0   
8696               NaN            MANUAL   rear wheel drive              2.0   
8697               NaN            MANUAL   rear wheel drive              2.0   
8698               NaN            MANUAL   rear wheel drive              4.0   
8699               NaN         AUTOMATIC   rear wheel drive              4.0   
8700               NaN            MANUAL   rear wheel drive              4.0   
8701               NaN            MANUAL   rear wheel drive              4.0   
8702               NaN            MANUAL   rear wheel drive              4.0   
8703               NaN         AUTOMATIC   rear wheel drive              4.0   
8704               NaN         AUTOMATIC   rear wheel drive              4.0   
8705  

In [77]:
df_no_duplicates[df_no_duplicates["engine_cylinders"] == 0].groupby(["engine_fuel_type"]).count()

make  model  year  engine_hp  engine_cylinders  \
engine_fuel_type                                                   
electric            56     56    56         13                56   

                  transmission_type  driven_wheels  number_of_doors  \
engine_fuel_type                                                      
electric                         56             56               51   

                  market_category  vehicle_size  vehicle_style  highway_mpg  \
engine_fuel_type                                                              
electric                       51            56             56           56   

                  city_mpg  popularity  msrp  
engine_fuel_type                              
electric                56          56    56

In [78]:
df_no_duplicates[df_no_duplicates["engine_cylinders"].isna()].groupby(["engine_fuel_type"]).count()

make  model  year  engine_hp  engine_cylinders  \
engine_fuel_type                                                              
electric                       10     10    10          9                 0   
premium unleaded (required)    17     17    17         17                 0   
regular unleaded                3      3     3          3                 0   

                             transmission_type  driven_wheels  \
engine_fuel_type                                                
electric                                    10             10   
premium unleaded (required)                 17             17   
regular unleaded                             3              3   

                             number_of_doors  market_category  vehicle_size  \
engine_fuel_type                                                              
electric                                  10               10            10   
premium unleaded (required)               17               17            17   
regular unleaded                           3                3             3   

                             vehicle_style  highway_mpg  city_mpg  popularity  \
engine_fuel_type                                                                
electric                                10           10        10          10   
premium unleaded (required)             17           17        17          17   
regular unleaded                         3            3         3           3   

                             msrp  
engine_fuel_type                   
electric                       10  
premium unleaded (required)    17  
regular unleaded                3

## I decided to transform the engine_cylinders from the electric vehicles to NaN

In [79]:
df_no_duplicates.loc[df_no_duplicates["engine_cylinders"] == 0, "engine_cylinders"] = None

In [80]:
df_no_duplicates[df_no_duplicates["engine_cylinders"] == 0]

Empty DataFrame
Columns: [make, model, year, engine_fuel_type, engine_hp, engine_cylinders, transmission_type, driven_wheels, number_of_doors, market_category, vehicle_size, vehicle_style, highway_mpg, city_mpg, popularity, msrp]
Index: []

## fill the missing values on the column "engine_fuel_type"

In [81]:
df_no_duplicates[df_no_duplicates["engine_fuel_type"].isna()]

make   model  year engine_fuel_type  engine_hp  engine_cylinders  \
11321  Suzuki  Verona  2004              NaN      155.0               6.0   
11322  Suzuki  Verona  2004              NaN      155.0               6.0   
11323  Suzuki  Verona  2004              NaN      155.0               6.0   

      transmission_type      driven_wheels  number_of_doors market_category  \
11321         AUTOMATIC  front wheel drive              4.0             NaN   
11322         AUTOMATIC  front wheel drive              4.0             NaN   
11323         AUTOMATIC  front wheel drive              4.0             NaN   

      vehicle_size vehicle_style  highway_mpg  city_mpg  popularity   msrp  
11321      Midsize         Sedan           25        17         481  17199  
11322      Midsize         Sedan           25        17         481  20199  
11323      Midsize         Sedan           25        17         481  18499

In [82]:
df_no_duplicates[df_no_duplicates["model"] == "Verona"]

make   model  year  engine_fuel_type  engine_hp  engine_cylinders  \
11321  Suzuki  Verona  2004               NaN      155.0               6.0   
11322  Suzuki  Verona  2004               NaN      155.0               6.0   
11323  Suzuki  Verona  2004               NaN      155.0               6.0   
11324  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11325  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11326  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11327  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11328  Suzuki  Verona  2006  regular unleaded      155.0               6.0   
11329  Suzuki  Verona  2006  regular unleaded      155.0               6.0   

      transmission_type      driven_wheels  number_of_doors market_category  \
11321         AUTOMATIC  front wheel drive              4.0             NaN   
11322         AUTOMATIC  front wheel drive              4.0             NaN   
11323         AUTOMATIC  front wheel drive              4.0             NaN   
11324         AUTOMATIC  front wheel drive              4.0             NaN   
11325         AUTOMATIC  front wheel drive              4.0             NaN   
11326         AUTOMATIC  front wheel drive              4.0             NaN   
11327         AUTOMATIC  front wheel drive              4.0             NaN   
11328         AUTOMATIC  front wheel drive              4.0             NaN   
11329         AUTOMATIC  front wheel drive              4.0             NaN   

      vehicle_size vehicle_style  highway_mpg  city_mpg  popularity   msrp  
11321      Midsize         Sedan           25        17         481  17199  
11322      Midsize         Sedan           25        17         481  20199  
11323      Midsize         Sedan           25        17         481  18499  
11324      Midsize         Sedan           25        18         481  19349  
11325      Midsize         Sedan           25        18         481  21049  
11326      Midsize         Sedan           25        18         481  17549  
11327      Midsize         Sedan           25        18         481  20549  
11328      Midsize         Sedan           25        17         481  20299  
11329      Midsize         Sedan           25        17         481  18299

### setting the values to regular unleaded, imputate the missing values through infos of the dataset and domain knowledge

In [83]:
df_no_duplicates.loc[df_no_duplicates["engine_fuel_type"].isna(), "engine_fuel_type"] = "regular unleaded"
df_no_duplicates[df_no_duplicates["model"] == "Verona"]

make   model  year  engine_fuel_type  engine_hp  engine_cylinders  \
11321  Suzuki  Verona  2004  regular unleaded      155.0               6.0   
11322  Suzuki  Verona  2004  regular unleaded      155.0               6.0   
11323  Suzuki  Verona  2004  regular unleaded      155.0               6.0   
11324  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11325  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11326  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11327  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11328  Suzuki  Verona  2006  regular unleaded      155.0               6.0   
11329  Suzuki  Verona  2006  regular unleaded      155.0               6.0   

      transmission_type      driven_wheels  number_of_doors market_category  \
11321         AUTOMATIC  front wheel drive              4.0             NaN   
11322         AUTOMATIC  front wheel drive              4.0             NaN   
11323         AUTOMATIC  front wheel drive              4.0             NaN   
11324         AUTOMATIC  front wheel drive              4.0             NaN   
11325         AUTOMATIC  front wheel drive              4.0             NaN   
11326         AUTOMATIC  front wheel drive              4.0             NaN   
11327         AUTOMATIC  front wheel drive              4.0             NaN   
11328         AUTOMATIC  front wheel drive              4.0             NaN   
11329         AUTOMATIC  front wheel drive              4.0             NaN   

      vehicle_size vehicle_style  highway_mpg  city_mpg  popularity   msrp  
11321      Midsize         Sedan           25        17         481  17199  
11322      Midsize         Sedan           25        17         481  20199  
11323      Midsize         Sedan           25        17         481  18499  
11324      Midsize         Sedan           25        18         481  19349  
11325      Midsize         Sedan           25        18         481  21049  
11326      Midsize         Sedan           25        18         481  17549  
11327      Midsize         Sedan           25        18         481  20549  
11328      Midsize         Sedan           25        17         481  20299  
11329      Midsize         Sedan           25        17         481  18299

## fill missing values of column "number of doors"

In [84]:
df_no_duplicates[df_no_duplicates["number_of_doors"].isna()]

make    model  year             engine_fuel_type  engine_hp  \
4666  Ferrari       FF  2013  premium unleaded (required)      651.0   
6930    Tesla  Model S  2016                     electric        NaN   
6931    Tesla  Model S  2016                     electric        NaN   
6932    Tesla  Model S  2016                     electric        NaN   
6933    Tesla  Model S  2016                     electric        NaN   
6934    Tesla  Model S  2016                     electric        NaN   

      engine_cylinders transmission_type     driven_wheels  number_of_doors  \
4666              12.0  AUTOMATED_MANUAL   all wheel drive              NaN   
6930               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6931               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6932               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6933               NaN      DIRECT_DRIVE  rear wheel drive              NaN   
6934               NaN      DIRECT_DRIVE   all wheel drive              NaN   

              market_category vehicle_size vehicle_style  highway_mpg  \
4666  Exotic,High-Performance        Large         Coupe           16   
6930       Exotic,Performance        Large         Sedan          105   
6931       Exotic,Performance        Large         Sedan          101   
6932  Exotic,High-Performance        Large         Sedan          105   
6933       Exotic,Performance        Large         Sedan          100   
6934       Exotic,Performance        Large         Sedan          107   

      city_mpg  popularity    msrp  
4666        11        2774  295000  
6930       102        1391   79500  
6931        98        1391   66000  
6932        92        1391  134500  
6933        97        1391   74500  
6934       101        1391   71000

In [85]:
df_no_duplicates[(df_no_duplicates["model"] == "FF") | (df_no_duplicates["model"] == "Model S")]

make    model  year             engine_fuel_type  engine_hp  \
4666  Ferrari       FF  2013  premium unleaded (required)      651.0   
4667  Ferrari       FF  2014  premium unleaded (required)      651.0   
4668  Ferrari       FF  2015  premium unleaded (required)      651.0   
6921    Tesla  Model S  2014                     electric        NaN   
6922    Tesla  Model S  2014                     electric        NaN   
6923    Tesla  Model S  2014                     electric        NaN   
6924    Tesla  Model S  2014                     electric        NaN   
6925    Tesla  Model S  2015                     electric        NaN   
6926    Tesla  Model S  2015                     electric        NaN   
6927    Tesla  Model S  2015                     electric        NaN   
6928    Tesla  Model S  2015                     electric        NaN   
6929    Tesla  Model S  2015                     electric        NaN   
6930    Tesla  Model S  2016                     electric        NaN   
6931    Tesla  Model S  2016                     electric        NaN   
6932    Tesla  Model S  2016                     electric        NaN   
6933    Tesla  Model S  2016                     electric        NaN   
6934    Tesla  Model S  2016                     electric        NaN   
6935    Tesla  Model S  2016                     electric        NaN   
6936    Tesla  Model S  2016                     electric        NaN   
6937    Tesla  Model S  2016                     electric        NaN   
6938    Tesla  Model S  2016                     electric        NaN   

      engine_cylinders transmission_type     driven_wheels  number_of_doors  \
4666              12.0  AUTOMATED_MANUAL   all wheel drive              NaN   
4667              12.0  AUTOMATED_MANUAL   all wheel drive              2.0   
4668              12.0  AUTOMATED_MANUAL   all wheel drive              2.0   
6921               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6922               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6923               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6924               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6925               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6926               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6927               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6928               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6929               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6930               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6931               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6932               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6933               NaN      DIRECT_DRIVE  rear wheel drive              NaN   
6934               NaN      DIRECT_DRIVE   all wheel drive              NaN   
6935               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6936               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6937               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6938               NaN      DIRECT_DRIVE  rear wheel drive              4.0   

              market_category vehicle_size vehicle_style  highway_mpg  \
4666  Exotic,High-Performance        Large         Coupe           16   
4667  Exotic,High-Performance        Large         Coupe           16   
4668  Exotic,High-Performance        Large         Coupe           16   
6921  Exotic,High-Performance        Large         Sedan           90   
6922       Exotic,Performance        Large         Sedan           97   
6923  Exotic,High-Performance        Large         Sedan           94   
6924  Exotic,High-Performance        Large         Sedan           90   
6925       Exotic,Performance        Large         Sedan           97   
6926       Exotic,Performanc

In [86]:
df_no_duplicates.loc[
    (df_no_duplicates["model"] == "Model S") & (df_no_duplicates["number_of_doors"].isna()), 
    "number_of_doors"
] = 4


In [87]:
df_no_duplicates.loc[
    (df_no_duplicates["model"] == "FF") & (df_no_duplicates["number_of_doors"].isna()),
    "number_of_doors"
    ] = 2

In [88]:
df_no_duplicates[(df_no_duplicates["model"] == "FF") | (df_no_duplicates["model"] == "Model S")]

make    model  year             engine_fuel_type  engine_hp  \
4666  Ferrari       FF  2013  premium unleaded (required)      651.0   
4667  Ferrari       FF  2014  premium unleaded (required)      651.0   
4668  Ferrari       FF  2015  premium unleaded (required)      651.0   
6921    Tesla  Model S  2014                     electric        NaN   
6922    Tesla  Model S  2014                     electric        NaN   
6923    Tesla  Model S  2014                     electric        NaN   
6924    Tesla  Model S  2014                     electric        NaN   
6925    Tesla  Model S  2015                     electric        NaN   
6926    Tesla  Model S  2015                     electric        NaN   
6927    Tesla  Model S  2015                     electric        NaN   
6928    Tesla  Model S  2015                     electric        NaN   
6929    Tesla  Model S  2015                     electric        NaN   
6930    Tesla  Model S  2016                     electric        NaN   
6931    Tesla  Model S  2016                     electric        NaN   
6932    Tesla  Model S  2016                     electric        NaN   
6933    Tesla  Model S  2016                     electric        NaN   
6934    Tesla  Model S  2016                     electric        NaN   
6935    Tesla  Model S  2016                     electric        NaN   
6936    Tesla  Model S  2016                     electric        NaN   
6937    Tesla  Model S  2016                     electric        NaN   
6938    Tesla  Model S  2016                     electric        NaN   

      engine_cylinders transmission_type     driven_wheels  number_of_doors  \
4666              12.0  AUTOMATED_MANUAL   all wheel drive              2.0   
4667              12.0  AUTOMATED_MANUAL   all wheel drive              2.0   
4668              12.0  AUTOMATED_MANUAL   all wheel drive              2.0   
6921               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6922               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6923               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6924               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6925               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6926               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6927               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6928               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6929               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6930               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6931               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6932               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6933               NaN      DIRECT_DRIVE  rear wheel drive              4.0   
6934               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6935               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6936               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6937               NaN      DIRECT_DRIVE   all wheel drive              4.0   
6938               NaN      DIRECT_DRIVE  rear wheel drive              4.0   

              market_category vehicle_size vehicle_style  highway_mpg  \
4666  Exotic,High-Performance        Large         Coupe           16   
4667  Exotic,High-Performance        Large         Coupe           16   
4668  Exotic,High-Performance        Large         Coupe           16   
6921  Exotic,High-Performance        Large         Sedan           90   
6922       Exotic,Performance        Large         Sedan           97   
6923  Exotic,High-Performance        Large         Sedan           94   
6924  Exotic,High-Performance        Large         Sedan           90   
6925       Exotic,Performance        Large         Sedan           97   
6926       Exotic,Performanc